geode-for-redis/src/distributedTest/java/org/apache/geode/redis/internal/commands/executor/list/LPushDUnitTest.java

```diff
@@ -0,0 +1,216 @@
+    jedis = new JedisCluster(new HostAndPort(BIND_ADDRESS, redisServerPort), REDIS_CLIENT_TIMEOUT);
+    clusterStartUp.flushAll();
+  }
+
+  @AfterClass
+  public static void tearDown() {
+    jedis.close();
+  }
+
+  @Test
+  public void shouldPushMultipleElementsAtomically()
+      throws ExecutionException, InterruptedException {
+    AtomicLong runningCount = new AtomicLong(PUSHER_COUNT);
+
+    List<String> listHashtags = makeListHashtags();
+    List<String> keys = makeListKeys(listHashtags);
+    List<String> elements1 = makeElementList(PUSH_LIST_SIZE, "element1-");
+    List<String> elements2 = makeElementList(PUSH_LIST_SIZE, "element2-");
+    List<String> elements3 = makeElementList(PUSH_LIST_SIZE, "element3-");
+
+    List<Runnable> taskList = new ArrayList<>();
+    taskList.add(() -> lpushPerformAndVerify(keys.get(0), elements1, runningCount));
+    taskList.add(() -> lpushPerformAndVerify(keys.get(0), elements1, runningCount));
+    taskList.add(() -> lpushPerformAndVerify(keys.get(1), elements2, runningCount));
+    taskList.add(() -> lpushPerformAndVerify(keys.get(1), elements2, runningCount));
+    taskList.add(() -> lpushPerformAndVerify(keys.get(2), elements3, runningCount));
+    taskList.add(() -> lpushPerformAndVerify(keys.get(2), elements3, runningCount));
+    taskList.add(() -> verifyListLengthCondition(keys.get(0), runningCount));
+    taskList.add(() -> verifyListLengthCondition(keys.get(1), runningCount));
+    taskList.add(() -> verifyListLengthCondition(keys.get(2), runningCount));
+
+    List<Future> futureList = new ArrayList<>();
+    for (Runnable task : taskList) {
+      futureList.add(executor.runAsync(task));
+    }
+
+    for (int i = 0; i < 50 && runningCount.get() > 0; i++) {
+      clusterStartUp.moveBucketForKey(listHashtags.get(i % listHashtags.size()));
+      GeodeAwaitility.await().during(Duration.ofMillis(500)).until(() -> true);
+    }
+
+    for (Future future : futureList) {
+      future.get();
+    }
+
+    int totalLength = 0;
+    Long length;
+    for (String key : keys) {
+      length = jedis.llen(key);
+      assertThat(length).isGreaterThanOrEqualTo(MINIMUM_ITERATIONS * 2 * PUSH_LIST_SIZE);
+      totalLength += length;
+    }
+    assertThat(totalLength)
+        .isGreaterThanOrEqualTo(MINIMUM_ITERATIONS * PUSHER_COUNT * PUSH_LIST_SIZE);
+    clusterStartUp.crashVM(1); // kill primary server, just in case test order is reversed
+  }
+
+  private void lpushPerformAndVerify(String key, List<String> elementList,
+      AtomicLong runningCount) {
+    for (int i = 0; i < MINIMUM_ITERATIONS; i++) {
+      long listLength = jedis.llen(key);
+      long newLength = jedis.lpush(key, elementList.toArray(new String[] {}));
+      assertThat((newLength - listLength) % 3).as("LPUSH, list length %s not multiple of 3",
+          newLength).isEqualTo(0);
+    }
+    runningCount.decrementAndGet();
+  }
+
+  @Test
+  public void shouldDistributeElementsAcrossCluster()
```

Comment:
This test would be better named something like "shouldNotLoseData_givenPrimaryServerCrashesDuringOperations" since that's what we're really trying to test here.

Reply:
Renamed.